### PR TITLE
Site filter performance test using released data

### PR DESCRIPTION
**Pull request by leehart**
Task https://quire.io/w/malariagen-ag3.0-paper/87

### CODE REVIEW

**Comment by leehart** [comment]
Currently takes 6.5 hours to run. (No assistance from KubeCluster or Numba.)

Means with SD for all crosses and chroms (training sites excluded).
image

Same with training and Ag2 crosses removed only seems to make a slight difference to the error bars:
image

Notebook also currently includes example plot of metrics {TPR, FPR, J} for crosses {73-2, 78-3} and chroms {2L, X}:
image

**Review by hardingnj — COMMENTED** [review]
Thanks for this Lee. I’ve been through the code, it makes sense- I can follow the logic of it. I don’t see anything that worries me.

I do think though that we can engineer this a little better- I worry that with this type of procedural code it’s harder to maintain, and lots of repeated similar variable names can allow bugs to creep in. 

Fundamentally, the code should be structured around 2 arrays. The ground truth array, and the corresponding array that we test this against. This is almost the case here, but I think it’s buried a bit too deep in intermediate steps. 

The key function in this code `calc_error_stats`, takes 2 arrays (int and bool). The equivalent function in my code is `cross_tab_values` at (https://github.com/malariagen/vector-ops/blob/master/work/1069-accessibility-base-filters/gamb_colu/accessibility_performance_assessment.ipynb). Both seem to have the same inputs and expected output. This function is a good example of something that I think could be simplified. Essentially, we have 4 categories that the result can be, or in the event of a -1 in the ground truth, we skip. The loop in `cross_tab_values` only visits each site once, and performs 3 tests for equality (and 1 test if -1). I count an additional 3 array wide operations in the `calc_error_stats` function. This function can be compiled using numba and applied blockwise over our arrays which will result in significant speed up. Then we can just stack up the resulting 2x2 arrays and take the sum, to give us our FP/TP/FN/TN values.

At the moment, the code slices out the training sites. I don’t think this is necessary. We could use `da.where` to flip these sites to -1 in the ground truth. This adds lots of complexity in checking the resulting sizes, and means lots of underlying data has to be shuffled around.  

Generally the solution as implemented has an inclination to grab pointers to all data and intermediates and stores in dictionaries- then operate on each dictionary at each step. I think all intermediates could be processed within a single (nested) loop.  

Some pseudocode:
```
for chrom in chromosomes:
    # load phase 3 mask
    p3_mask = v3.load_mask(chrom)

    # load phase 2 mask
    p2_mask = v2.load_mask(chrom)
    
    #  load_training mask
    training_mask = v3.load_mask(field="")
    for cross_id in crosses:
         ground_truth = load_ground_truth(chrom, cross_id)
         training_masked_ground_truth = da.where(~training_mask, ground_truth, -1)
     
         # these can be sped up using map_blocks
         p3_results = calc_error_stats(training_masked, p3_mask)

         p2_results = calc_error_stats(training_masked, p2_mask)
         # all pass results (permissive) are trivial, as we just move all TNs to FPs, FNs to TPs.  
            
         # above to be stored in a dict or `pd.DataFrame`
```

I think the additional code in ag3.py, and ag2.py are great- these are nice additions that simplify the code.

Happy to discuss further on Monday

**Comment by leehart** [comment]
Thanks for the review Nick.

Some of what you're suggesting sounds like coding style preference to me, with different values traded off. I might personally find this style of procedural code easier maintain, easier to follow, easier to reason about, and vice versa. But there is obviously some objective room for improvement, particularly in the metrics of speed and efficiency, which should be achievable without obscuring the data narrative too much - I'm conscious of that and I'm still looking into it. I took your original code and re-wrote it in a way that I could easily follow, mindful that others would need to follow it too, sacrificing those performance elements until I understood them better.

For instance, I would worry that using `da.where()` to flip the training sites to `-1` in the ground truth arrays would subvert the original semantics of `-1`, which indicates that a site's good/bad status is undetermined, wrt Mendelian error or male heterozygosity. Although that might be an efficient performance hack, and it might simplify/shorten some code downstream, it might also make the data provenance harder to track, and the code narrative harder to follow, and could actually create more complexity/confusion in those areas. Of course, those perceptions are subjective, and a simple code comment to explain would probably suffice!

Having said that, of course I'm trying to see it all from your point of view, and take everything on board, and hopefully I can synthesize something that appeals to both our tastes, both aesthetically and technically, and then we should land on common ground, in a better place. :-)

**Comment by hardingnj** [comment]
Cheers Lee. Thinking about it, I think we could even include the training set mask in the function that generates the ground truth. I think it would fit there. 

It is definitely a style thing- I don't think there are any issues with the content of the code. I do feel though that its always better to do all the intermediate steps/processing together so the logical flow is clearer. I find it hard to follow all the `dict()` objects floating around.

**Comment by leehart** [comment]
> I find it hard to follow all the `dict()` objects floating around.

Too Many Dicts On The Dance Floor - Flight Of The Conchords

**Comment by leehart** [comment]
@hardingnj I've added a version of the notebook (and its output) that is simply a copy of your [vector-ops one](https://github.com/malariagen/vector-ops/blob/master/work/1069-accessibility-base-filters/gamb_colu/accessibility_performance_assessment.ipynb) with minimal changes: it sources the publicly released data (via included the Python modules); it only compares Ag3 against phase 2 and 100% pass; and it excludes the training data using your `da.where` suggestion. On a user pod, it runs about 8 mins faster than my dict-centric-with-intermediates version. I also removed the cluster dependency, just to simplify. Let me know if that was important to leave in.

image

I expect that's what you wanted, but I still find that approach too cryptic and convoluted. It's funny because you have a similar impression of my version, which I tried to make more step-by-step and easier to follow! I'm very reluctant to over-engineer this, especially if it costs readability. This is just a little script that fetches some data (it's not like another multinational e-commerce CMS) but I still value quality and craftsmanship, obviously! This seems like an awkward forum to be debating software engineering principles.

Again, this seems to raise the question: what is this repo for? If we only need to produce a table and perhaps a plot for the paper, then both these notebooks can do that, within the same time frame. Both versions also provide examples of how to access the publicly released data.

If we also need to provide examples of well-engineered and highly-readable notebooks, to show how the CSVs and SVGs were compiled, and to demonstrate best practices, then this is likely to be more subjective, even though we both have the same objective. I would personally lean towards readability and lean away from over-engineering, but clearly different people have different takes on the same object. If I'm asked to paint the picture, then we'll get it one way -- and if you're asked to paint the picture, then we'll get it another way, although we're both trying to be "synergistic". I don't know how important it is to nail this thing, but I expect consensus would help steer us through the middle... but then there's [design by committee](https://en.wikipedia.org/wiki/Design_by_committee)! (-:

**Comment by hardingnj** [comment]
Thanks Lee. Appreciate the effort that you have put into this. 

> I also removed the cluster dependency, just to simplify. Let me know if that was important to leave in

I think that's fine. If it's only 45 minutes- that's a reasonable simplification I think. 

> On a user pod, it runs about 8 mins faster than my dict-centric-with-intermediates version. 

Well done on getting that time down in your version- it was 6 hours I think you said above, almost a 6x speed up!

> it excludes the training data using your da.where suggestion

Cool- thanks. That's certainly a very welcome and not straightforward addition.

I guess the main thing is that this repo is:
a) correct
b) clear

I still feel that someone coming in cold would be able to more easily follow this version- but I appreciate this is subjective to an extent. I'm happy that we both arrived at the same result though. I don't think there is necessarily harm in softening our positions a bit. I recognise that I could be more explicit in my coding, and it's good to challenge that.

**Comment by leehart** [comment]
Thanks Nick.

I reckon 6.5 hours to 45 minutes is about 8.7 times faster! Numba FTW

I found your `da.where` suggestion to be a straightforward addition to your code, but with the aforementioned concern about corrupting semantics. Maybe I did something wrong, although the results seem OK. All the amends to that notebook seemed quick and straightforward, actually, because I wanted to keep changes to a minimum, to preserve provenance.
- Main changes https://app.reviewnb.com/malariagen/ag1000g-phase3-data-paper/commit/779e80322f2b33747d8329ed211b55ddcd26a25d/
- Fix error bars https://app.reviewnb.com/malariagen/ag1000g-phase3-data-paper/commit/cf2d9915287935342161697d2cd4496bd91d8ba1/
- Exclude training sites (which you can see here doesn't actually make much difference to the mean values) https://app.reviewnb.com/malariagen/ag1000g-phase3-data-paper/commit/63a96e213e680ea45bb358f156b3e5ad075ce0c1/

Yes, I want to find ways to learn and appreciate the methods and benefits of doing all the intermediate steps together. That seems to be your fondness! And I need to see the problems with my own inclinations, using beaucoupe de dicts, etc.

I'm still learning Python. I'm not sure the `ag3.py` and `ag2.py` modules are properly engineered. The import dance seems a bit clunky for some reason. I tried to keep `ag2.py` consistent with the existing `ag3.py` at least.

**Review by hardingnj — APPROVED** [review]
LGTM- thanks Lee.
